### DISCUSSION TITLE
How to debug module in VS if running Oqtane in IIS (e.g. attach to process)?

### DISCUSSION THREAD

**Discussion by lanthonyneville** [oqtane/oqtane.framework#2576] [Q&A] [answered]
I know that one way to do module development is to open the Oqtane framework solution in Visual Studio, then add the projects for your module to the solution, and launch the whole application from VS.

However, I prefer to run my development instance of [Oqtane under IIS](https://www.oqtane.org/blog/!/24/installing-oqtane-on-iis), and maintain a VS solution that just contains the module projects, and copy the DLLs and wwwroot content into the instance on build  (and restart the IIS app pool). The reason for this approach is to prevent developers straying outside the module code and altering the Oqtane code; also its easier to see and enforce that the module is built against a specific Oqtane version (i.e. NuGet package).

So this approach does work, except I can't see how to debug my module code when the app is running, but this could be because I am new to Blazor apps. If this was a WebForms or MVC application, I would use the VS "attach to process" command to attach to w3wp.exe. However, Blazor apps running under IIS clearly don't work like that.

Is this kind of approach possible, and how?

Thanks

**Comment by leigh-pointer** [comment]
I have done this in the past! What is really important is setting up the build and xcopy paths in the debug.cmd plus set the server exe as your startup and then no need to attach.
________________________________
From: lanthonyneville ***@***.***>
Sent: Wednesday, February 1, 2023 4:08:12 PM
To: oqtane/oqtane.framework ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [oqtane/oqtane.framework] How to debug module in VS if running Oqtane in IIS (e.g. attach to process)? (Discussion #2576)


I know that one way to do module development is to open the Oqtane framework solution in Visual Studio, then add the projects for your module to the solution, and launch the whole application from VS.

However, I prefer to run my development instance of Oqtane under IIS<https://docs.oqtane.org/articles/documentation/run-in-iis.html>, and maintain a VS solution that just contains the module projects, and copy the DLLs and wwwroot content into the instance on build (and restart the IIS app pool). The reason for this approach is to prevent developers straying outside the module code and altering the Oqtane code; also its easier to see and enforce that the module is built against a specific Oqtane version (i.e. NuGet package).

So this approach does work, except I can't see how to debug my module code when the app is running, but this could be because I am new to Blazor apps. If this was a WebForms or MVC application, I would use the VS "attach to process" command to attach to w3wp.exe. However, Blazor apps running under IIS clearly don't work like that.

Is this kind of approach possible, and how?

Thanks

**Reply by lanthonyneville** [reply]
What do you mean "set the server exe as your startup"? Can you spell that out for me, sorry. (I have the debug.cmd set up to do the copy ... that bit is fine).

**Comment by leigh-pointer** [comment]
From your server module project properties, set the oqtane server exe as startup. The module server project should be set as start from the solution explorer.
________________________________

**Reply by lanthonyneville** [reply]
Thanks @leigh-pointer  .... I eventually found it ... in VS 2019 it was under Web in the Project Properties, but in VS 2022 you have to go to Project Properties > Debug > General > then open the Launch Profiles window and create a new profile that runs the Oqtane.Server.exe

However, I still think I am missing something with this because starting the server project opens a command window in which Oqtane.Server.exe seems to be running, but it throws several errors.

In the end I found out that attaching to w3wp.exe can be done .... I'll post that below ....

**Comment by lanthonyneville** [comment, marked as answer]
It seems I was mistaken about not being able to use Attach to Process to attach to w3wp.exe (due to some misconfiguration, the process was not showing up to attach to).

If you make the following selections on the Attach to Process window, you should find that the w3wp.exe process shows up and can be attached to, which means you can debug your module assembly code:

- Connection type: Default
- Connection target: your machine
- Attach to: Managed (.NET Core, .NET 5+) code
- Show processes for all users: checked
